swarm/network/fetcher.go

```diff
@@ -49,7 +52,29 @@ type Request struct {
 	Addr        storage.Address  // chunk address
 	Source      *discover.NodeID // nodeID of peer to request from (can be nil)
 	SkipCheck   bool             // whether to offer the chunk first or deliver directly
-	PeersToSkip *sync.Map        // peers not to request chunk from (only makes sense if source is nil)
+	peersToSkip *sync.Map        // peers not to request chunk from (only makes sense if source is nil)
+}
+
+func NewRequest(addr storage.Address, skipCheck bool, peersToSkip *sync.Map) *Request {
+	return &Request{
+		Addr:        addr,
+		SkipCheck:   skipCheck,
+		peersToSkip: peersToSkip,
+	}
+}
+
+func (r *Request) SkipPeer(nodeID string) bool {
+	val, ok := r.peersToSkip.Load(nodeID)
+	if !ok {
+		return false
+	}
+	t, ok := val.(time.Time)
+	if ok && time.Now().After(t.Add(RequestTimeout)) {
+		// deadine expired
+		r.peersToSkip.Delete(nodeID)
+		return false
+	}
+	return true
```

Comment:
lets create an issue to add unit tests for this

Reply:
Created https://github.com/ethersphere/go-ethereum/issues/921.